File: packages/stim-proto/src/content.ts
export interface LayoutHint {
  layout_family?: string;
  min_height_px?: number;
  max_height_px?: number;
  vertical_pressure?: string;
  metadata?: Record<string, unknown>;
}

export interface MessageContent {
  parts: ContentPart[];
  layout_hint?: LayoutHint;
}

interface BaseContentPart {
  part_id: string;
  revision: number;
  metadata?: Record<string, unknown>;
}

export interface TextPart extends BaseContentPart {
  type: "text";
  text: string;
}

export interface AssetRefPart extends BaseContentPart {
  type: "asset_ref";
  asset_ref: string;
  mime_type?: string;
}

export interface StimDomFragmentPayload {
  format: "stim-dom-fragment/v1";
  tree: Record<string, unknown>;
  bindings?: Record<string, unknown>;
}

export interface RawHtmlDomFragmentPayload {
  format: "raw_html";
  html: string;
  bindings?: Record<string, unknown>;
}

export interface DomFragmentPart extends BaseContentPart {
  type: "dom_fragment";
  payload: StimDomFragmentPayload | RawHtmlDomFragmentPayload;
}

export interface CapabilityRefPart extends BaseContentPart {
  type: "capability_ref";
  capability_kind: string;
  capability_id: string;
  input_schema_ref?: string;
  resource_ref?: string;
}

export type ContentPart =
  | TextPart
  | AssetRefPart
  | DomFragmentPart
  | CapabilityRefPart;

export interface PatchOperation {
  index: number;
  merge: Record<string, unknown>;
}

export interface InsertOperation {
  index: number;
  part: ContentPart;
}

File: crates/stim-proto/src/content.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use utoipa::ToSchema;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct LayoutHint {
    pub layout_family: Option<String>,
    pub min_height_px: Option<u32>,
    pub max_height_px: Option<u32>,
    pub vertical_pressure: Option<String>,
    pub metadata: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct MessageContent {
    pub parts: Vec<ContentPart>,
    pub layout_hint: Option<LayoutHint>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ContentPart {
    Text(TextPart),
    AssetRef(AssetRefPart),
    DomFragment(DomFragmentPart),
    CapabilityRef(CapabilityRefPart),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct TextPart {
    pub part_id: String,
    pub revision: u64,
    pub metadata: Option<Value>,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct AssetRefPart {
    pub part_id: String,
    pub revision: u64,
    pub metadata: Option<Value>,
    pub asset_ref: String,
    pub mime_type: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
#[serde(tag = "format", rename_all = "snake_case")]
pub enum DomFragmentPayload {
    #[serde(rename = "stim-dom-fragment/v1")]
    StimDomFragmentV1 {
        tree: Value,
        bindings: Option<Value>,
    },
    RawHtml {
        html: String,
        bindings: Option<Value>,
    },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct DomFragmentPart {
    pub part_id: String,
    pub revision: u64,
    pub metadata: Option<Value>,
    #[serde(flatten)]
    pub payload: DomFragmentPayload,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct CapabilityRefPart {
    pub part_id: String,
    pub revision: u64,
    pub metadata: Option<Value>,
    pub capability_kind: String,
    pub capability_id: String,
    pub input_schema_ref: Option<String>,
    pub resource_ref: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct PatchOperation {
    pub index: usize,
    pub merge: Value,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ToSchema)]
pub struct InsertOperation {
    pub index: usize,
    pub part: ContentPart,
}
